In [35]:
import pandas as pd
import numpy as np

# Project Details

## In this file task are:
- Clean Dataset
- Reducing size of dataset (if possible)

# 1.  Clean Dataset

In [54]:
# Transforming csv object to Data Frame
games = pd.read_csv('vgsales.csv')
# Creating copy 
games_copy = games.copy()

In [55]:
games_copy.head()

Rank                      Name Platform    Year         Genre Publisher  \
0     1                Wii Sports      Wii  2006.0        Sports  Nintendo   
1     2         Super Mario Bros.      NES  1985.0      Platform  Nintendo   
2     3            Mario Kart Wii      Wii  2008.0        Racing  Nintendo   
3     4         Wii Sports Resort      Wii  2009.0        Sports  Nintendo   
4     5  Pokemon Red/Pokemon Blue       GB  1996.0  Role-Playing  Nintendo   

   NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  
0     41.49     29.02      3.77         8.46         82.74  
1     29.08      3.58      6.81         0.77         40.24  
2     15.85     12.88      3.79         3.31         35.82  
3     15.75     11.01      3.28         2.96         33.00  
4     11.27      8.89     10.22         1.00         31.37

In [56]:
games_copy.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16598 entries, 0 to 16597
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Rank          16598 non-null  int64  
 1   Name          16598 non-null  object 
 2   Platform      16598 non-null  object 
 3   Year          16327 non-null  float64
 4   Genre         16598 non-null  object 
 5   Publisher     16540 non-null  object 
 6   NA_Sales      16598 non-null  float64
 7   EU_Sales      16598 non-null  float64
 8   JP_Sales      16598 non-null  float64
 9   Other_Sales   16598 non-null  float64
 10  Global_Sales  16598 non-null  float64
dtypes: float64(6), int64(1), object(4)
memory usage: 5.2 MB


In [57]:
games_copy.isnull().sum()

Rank              0
Name              0
Platform          0
Year            271
Genre             0
Publisher        58
NA_Sales          0
EU_Sales          0
JP_Sales          0
Other_Sales       0
Global_Sales      0
dtype: int64

In [58]:
# deleting NA values from Year column
games_copy.dropna(subset=['Year'],inplace=True)

In [59]:
games_copy.isnull().sum()

Rank             0
Name             0
Platform         0
Year             0
Genre            0
Publisher       36
NA_Sales         0
EU_Sales         0
JP_Sales         0
Other_Sales      0
Global_Sales     0
dtype: int64

In [60]:
# checking NaN values for publisher columns
games_copy.loc[games_copy.Publisher.isnull()]

Rank                                               Name Platform  \
1662    1664       Shrek / Shrek 2 2-in-1 Gameboy Advance Video      GBA   
2222    2224                                 Bentley's Hackpack      GBA   
3159    3161  Nicktoons Collection: Game Boy Advance Video V...      GBA   
3166    3168  SpongeBob SquarePants: Game Boy Advance Video ...      GBA   
3766    3768  SpongeBob SquarePants: Game Boy Advance Video ...      GBA   
4526    4528  The Fairly Odd Parents: Game Boy Advance Video...      GBA   
4635    4637  The Fairly Odd Parents: Game Boy Advance Video...      GBA   
5647    5649  Cartoon Network Collection: Game Boy Advance V...      GBA   
6437    6439           Sonic X: Game Boy Advance Video Volume 1      GBA   
6562    6564  Dora the Explorer: Game Boy Advance Video Volu...      GBA   
6648    6650  Cartoon Network Collection: Game Boy Advance V...      GBA   
6849    6851     All Grown Up!: Game Boy Advance Video Volume 1      GBA   
7208    7210  Nicktoons Collection: Game Boy Advance Video V...      GBA   
7470    7472  SpongeBob SquarePants: Game Boy Advance Video ...      GBA   
7953    7955                   Thomas the Tank Engine & Friends      GBA   
8330    8332    Dragon Ball GT: Game Boy Advance Video Volume 1      GBA   
8341    8343  Codename: Kids Next Door: Game Boy Advance Vid...      GBA   
8368    8370  Teenage Mutant Ninja Turtles: Game Boy Advance...      GBA   
8503    8505                                       Stronghold 3       PC   
8770    8772  Cartoon Network Collection: Game Boy Advance V...      GBA   
8848    8850  Pokémon: Johto Photo Finish: Game Boy Advance ...      GBA   
8896    8898  Strawberry Shortcake: Game Boy Advance Video V...      GBA   
9517    9519                             Farming Simulator 2011       PC   
10382  10384                   Disney Channel Collection Vol. 1      GBA   
11526  11528  Cartoon Network Collection: Game Boy Advance V...      GBA   
14311  14314                                      Silverlicious       DS   
14698  14701                                 UK Truck Simulator       PC   
15056  15059                                            Xia-Xia       DS   
15325  15328  Nicktoons Collection: Game Boy Advance Video V...      GBA   
15353  15356                   Demolition Company: Gold Edition       PC   
15788  15791  Moshi, Kono Sekai ni Kami-sama ga Iru to suru ...      PSV   
16208  16211  Super Robot Monkey Team: Game Boy Advance Vide...      GBA   
16367  16370                                  Dance with Devils      PSV   
16494  16497                    Legends of Oz: Dorothy's Return      3DS   
16543  16546                             Driving Simulator 2011       PC   
16553  16556                                     Bound By Flame     X360   

         Year         Genre Publisher  NA_Sales  EU_Sales  JP_Sales  \
1662   2007.0          Misc       NaN      0.87      0.32      0.00   
2222   2005.0          Misc       NaN      0.67      0.25      0.00   
3159   2004.0          Misc       NaN      0.46      0.17      0.00   
3166   2004.0          Misc       NaN      0.46      0.17      0.00   
3766   2004.0          Misc       NaN      0.38      0.14      0.00   
4526   2004.0          Misc       NaN      0.31      0.11      0.00   
4635   2004.0          Misc       NaN      0.30      0.11      0.00   
5647   2005.0          Misc       NaN      0.23      0.08      0.00   
6437   2004.0          Misc       NaN      0.19      0.07      0.00   
6562   2004.0          Misc       NaN      0.18      0.07      0.00   
6648   2004.0          Misc       NaN      0.18      0.07      0.00   
6849   2004.0          Misc       NaN      0.17      0.06      0.00   
7208   2004.0          Misc       NaN      0.16      0.06      0.00   
7470   2004.0          Misc       NaN      0.15      0.05      0.00   
7953   2004.0     Adventure       NaN      0.13      0.05      0.00   
8330   2004.0          Misc       NaN      0.12      0.05  

In [61]:
# changing Nan values
games_copy.Publisher.fillna('Unknown Publisher', inplace=True)

In [62]:
# changing index to title of the game
games_copy.set_index(games_copy.Name, inplace=True)

In [63]:
games_copy.head()

Rank                      Name Platform    Year  \
Name                                                                        
Wii Sports                   1                Wii Sports      Wii  2006.0   
Super Mario Bros.            2         Super Mario Bros.      NES  1985.0   
Mario Kart Wii               3            Mario Kart Wii      Wii  2008.0   
Wii Sports Resort            4         Wii Sports Resort      Wii  2009.0   
Pokemon Red/Pokemon Blue     5  Pokemon Red/Pokemon Blue       GB  1996.0   

                                 Genre Publisher  NA_Sales  EU_Sales  \
Name                                                                   
Wii Sports                      Sports  Nintendo     41.49     29.02   
Super Mario Bros.             Platform  Nintendo     29.08      3.58   
Mario Kart Wii                  Racing  Nintendo     15.85     12.88   
Wii Sports Resort               Sports  Nintendo     15.75     11.01   
Pokemon Red/Pokemon Blue  Role-Playing  Nintendo     11.27      8.89   

                          JP_Sales  Other_Sales  Global_Sales  
Name                                                           
Wii Sports                    3.77         8.46         82.74  
Super Mario Bros.             6.81         0.77         40.24  
Mario Kart Wii                3.79         3.31         35.82  
Wii Sports Resort             3.28         2.96         33.00  
Pokemon Red/Pokemon Blue     10.22         1.00         31.37

In [64]:
# Checking for duplicates in Data Frame
games_copy.loc[games_copy.duplicated()]

Empty DataFrame
Columns: [Rank, Name, Platform, Year, Genre, Publisher, NA_Sales, EU_Sales, JP_Sales, Other_Sales, Global_Sales]
Index: []

In [65]:
games_copy.describe()

Rank          Year      NA_Sales      EU_Sales      JP_Sales  \
count  16327.000000  16327.000000  16327.000000  16327.000000  16327.000000   
mean    8292.868194   2006.406443      0.265415      0.147554      0.078661   
std     4792.669778      5.828981      0.821591      0.508766      0.311557   
min        1.000000   1980.000000      0.000000      0.000000      0.000000   
25%     4136.500000   2003.000000      0.000000      0.000000      0.000000   
50%     8295.000000   2007.000000      0.080000      0.020000      0.000000   
75%    12441.500000   2010.000000      0.240000      0.110000      0.040000   
max    16600.000000   2020.000000     41.490000     29.020000     10.220000   

        Other_Sales  Global_Sales  
count  16327.000000  16327.000000  
mean       0.048325      0.540232  
std        0.189885      1.565732  
min        0.000000      0.010000  
25%        0.000000      0.060000  
50%        0.010000      0.170000  
75%        0.040000      0.480000  
max       10.570000     82.740000

In [66]:
games_copy.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
Index: 16327 entries, Wii Sports to Spirits & Spells
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Rank          16327 non-null  int64  
 1   Name          16327 non-null  object 
 2   Platform      16327 non-null  object 
 3   Year          16327 non-null  float64
 4   Genre         16327 non-null  object 
 5   Publisher     16327 non-null  object 
 6   NA_Sales      16327 non-null  float64
 7   EU_Sales      16327 non-null  float64
 8   JP_Sales      16327 non-null  float64
 9   Other_Sales   16327 non-null  float64
 10  Global_Sales  16327 non-null  float64
dtypes: float64(6), int64(1), object(4)
memory usage: 6.4 MB


# 2. Reducing size of Dataset

In [68]:
games_copy.Platform.unique()

array(['Wii', 'NES', 'GB', 'DS', 'X360', 'PS3', 'PS2', 'SNES', 'GBA',
       '3DS', 'PS4', 'N64', 'PS', 'XB', 'PC', '2600', 'PSP', 'XOne', 'GC',
       'WiiU', 'GEN', 'DC', 'PSV', 'SAT', 'SCD', 'WS', 'NG', 'TG16',
       '3DO', 'GG', 'PCFX'], dtype=object)

In [70]:
games_copy.Platform = games_copy.Platform.astype('category')

In [71]:
games_copy.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
Index: 16327 entries, Wii Sports to Spirits & Spells
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype   
---  ------        --------------  -----   
 0   Rank          16327 non-null  int64   
 1   Name          16327 non-null  object  
 2   Platform      16327 non-null  category
 3   Year          16327 non-null  float64 
 4   Genre         16327 non-null  object  
 5   Publisher     16327 non-null  object  
 6   NA_Sales      16327 non-null  float64 
 7   EU_Sales      16327 non-null  float64 
 8   JP_Sales      16327 non-null  float64 
 9   Other_Sales   16327 non-null  float64 
 10  Global_Sales  16327 non-null  float64 
dtypes: category(1), float64(6), int64(1), object(3)
memory usage: 5.8 MB


In [73]:
games_copy.Year = games_copy.Year.astype('int16')

In [83]:
# checking if changing Dtype reduced the size, if so continue with next column
games_copy.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
Index: 16327 entries, Wii Sports to Spirits & Spells
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype   
---  ------        --------------  -----   
 0   Rank          16327 non-null  int64   
 1   Name          16327 non-null  object  
 2   Platform      16327 non-null  category
 3   Year          16327 non-null  int16   
 4   Genre         16327 non-null  category
 5   Publisher     16327 non-null  category
 6   NA_Sales      16327 non-null  float64 
 7   EU_Sales      16327 non-null  float64 
 8   JP_Sales      16327 non-null  float64 
 9   Other_Sales   16327 non-null  float64 
 10  Global_Sales  16327 non-null  float64 
dtypes: category(3), float64(5), int16(1), int64(1), object(1)
memory usage: 3.7 MB


In [84]:
games_copy.Genre.unique()

['Sports', 'Platform', 'Racing', 'Role-Playing', 'Puzzle', ..., 'Simulation', 'Action', 'Fighting', 'Adventure', 'Strategy']
Length: 12
Categories (12, object): ['Sports', 'Platform', 'Racing', 'Role-Playing', ..., 'Action', 'Fighting', 'Adventure', 'Strategy']

In [77]:
games_copy.Genre = games_copy.Genre.astype('category')

In [79]:
games_copy.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
Index: 16327 entries, Wii Sports to Spirits & Spells
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype   
---  ------        --------------  -----   
 0   Rank          16327 non-null  int64   
 1   Name          16327 non-null  object  
 2   Platform      16327 non-null  category
 3   Year          16327 non-null  int16   
 4   Genre         16327 non-null  category
 5   Publisher     16327 non-null  object  
 6   NA_Sales      16327 non-null  float64 
 7   EU_Sales      16327 non-null  float64 
 8   JP_Sales      16327 non-null  float64 
 9   Other_Sales   16327 non-null  float64 
 10  Global_Sales  16327 non-null  float64 
dtypes: category(2), float64(5), int16(1), int64(1), object(2)
memory usage: 4.7 MB


In [81]:
games_copy.Publisher.unique()

array(['Nintendo', 'Microsoft Game Studios', 'Take-Two Interactive',
       'Sony Computer Entertainment', 'Activision', 'Ubisoft',
       'Bethesda Softworks', 'Electronic Arts', 'Sega', 'SquareSoft',
       'Atari', '505 Games', 'Capcom', 'GT Interactive',
       'Konami Digital Entertainment',
       'Sony Computer Entertainment Europe', 'Square Enix', 'LucasArts',
       'Virgin Interactive', 'Warner Bros. Interactive Entertainment',
       'Universal Interactive', 'Eidos Interactive', 'RedOctane',
       'Vivendi Games', 'Enix Corporation', 'Namco Bandai Games',
       'Palcom', 'Hasbro Interactive', 'THQ', 'Fox Interactive',
       'Acclaim Entertainment', 'MTV Games', 'Disney Interactive Studios',
       'Majesco Entertainment', 'Codemasters', 'Red Orb', 'Level 5',
       'Arena Entertainment', 'Midway Games', 'JVC', 'Deep Silver',
       '989 Studios', 'NCSoft', 'UEP Systems', 'Parker Bros.', 'Maxis',
       'Imagic', 'Tecmo Koei', 'Valve Software', 'ASCII Entertainment',
     

In [85]:
games_copy.Publisher = games_copy.Publisher.astype('category')

In [86]:
games_copy.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
Index: 16327 entries, Wii Sports to Spirits & Spells
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype   
---  ------        --------------  -----   
 0   Rank          16327 non-null  int64   
 1   Name          16327 non-null  object  
 2   Platform      16327 non-null  category
 3   Year          16327 non-null  int16   
 4   Genre         16327 non-null  category
 5   Publisher     16327 non-null  category
 6   NA_Sales      16327 non-null  float64 
 7   EU_Sales      16327 non-null  float64 
 8   JP_Sales      16327 non-null  float64 
 9   Other_Sales   16327 non-null  float64 
 10  Global_Sales  16327 non-null  float64 
dtypes: category(3), float64(5), int16(1), int64(1), object(1)
memory usage: 3.7 MB


In [88]:
games_copy.NA_Sales.mean()

0.2654149568200114

In [90]:
games_copy.NA_Sales.median()

0.08

with values that are not that big, we can try changing type to float16 to reduce the size even more

In [91]:
games_copy.NA_Sales = games_copy.NA_Sales.astype('float16')

In [92]:
games_copy.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
Index: 16327 entries, Wii Sports to Spirits & Spells
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype   
---  ------        --------------  -----   
 0   Rank          16327 non-null  int64   
 1   Name          16327 non-null  object  
 2   Platform      16327 non-null  category
 3   Year          16327 non-null  int16   
 4   Genre         16327 non-null  category
 5   Publisher     16327 non-null  category
 6   NA_Sales      16327 non-null  float16 
 7   EU_Sales      16327 non-null  float64 
 8   JP_Sales      16327 non-null  float64 
 9   Other_Sales   16327 non-null  float64 
 10  Global_Sales  16327 non-null  float64 
dtypes: category(3), float16(1), float64(4), int16(1), int64(1), object(1)
memory usage: 3.6 MB


the size of dataframe is a bit smaller, so will do the same for the rest

In [95]:
# checking first if the rest of values are close to NA_Sales
games_copy.describe()

Rank          Year      NA_Sales      EU_Sales      JP_Sales  \
count  16327.000000  16327.000000  16327.000000  16327.000000  16327.000000   
mean    8292.868194   2006.406443      0.265381      0.147554      0.078661   
std     4792.669778      5.828981      0.821289      0.508766      0.311557   
min        1.000000   1980.000000      0.000000      0.000000      0.000000   
25%     4136.500000   2003.000000      0.000000      0.000000      0.000000   
50%     8295.000000   2007.000000      0.080017      0.020000      0.000000   
75%    12441.500000   2010.000000      0.239990      0.110000      0.040000   
max    16600.000000   2020.000000     41.500000     29.020000     10.220000   

        Other_Sales  Global_Sales  
count  16327.000000  16327.000000  
mean       0.048325      0.540232  
std        0.189885      1.565732  
min        0.000000      0.010000  
25%        0.000000      0.060000  
50%        0.010000      0.170000  
75%        0.040000      0.480000  
max       10.570000     82.740000

In [96]:
games_copy.EU_Sales = games_copy.EU_Sales.astype('float16')

In [97]:
games_copy.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
Index: 16327 entries, Wii Sports to Spirits & Spells
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype   
---  ------        --------------  -----   
 0   Rank          16327 non-null  int64   
 1   Name          16327 non-null  object  
 2   Platform      16327 non-null  category
 3   Year          16327 non-null  int16   
 4   Genre         16327 non-null  category
 5   Publisher     16327 non-null  category
 6   NA_Sales      16327 non-null  float16 
 7   EU_Sales      16327 non-null  float16 
 8   JP_Sales      16327 non-null  float64 
 9   Other_Sales   16327 non-null  float64 
 10  Global_Sales  16327 non-null  float64 
dtypes: category(3), float16(2), float64(3), int16(1), int64(1), object(1)
memory usage: 3.5 MB


In [98]:
games_copy.JP_Sales = games_copy.JP_Sales.astype('float16')

In [99]:
games_copy.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
Index: 16327 entries, Wii Sports to Spirits & Spells
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype   
---  ------        --------------  -----   
 0   Rank          16327 non-null  int64   
 1   Name          16327 non-null  object  
 2   Platform      16327 non-null  category
 3   Year          16327 non-null  int16   
 4   Genre         16327 non-null  category
 5   Publisher     16327 non-null  category
 6   NA_Sales      16327 non-null  float16 
 7   EU_Sales      16327 non-null  float16 
 8   JP_Sales      16327 non-null  float16 
 9   Other_Sales   16327 non-null  float64 
 10  Global_Sales  16327 non-null  float64 
dtypes: category(3), float16(3), float64(2), int16(1), int64(1), object(1)
memory usage: 3.4 MB


In [100]:
games_copy.Other_Sales = games_copy.Other_Sales.astype('float16')

In [101]:
games_copy.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
Index: 16327 entries, Wii Sports to Spirits & Spells
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype   
---  ------        --------------  -----   
 0   Rank          16327 non-null  int64   
 1   Name          16327 non-null  object  
 2   Platform      16327 non-null  category
 3   Year          16327 non-null  int16   
 4   Genre         16327 non-null  category
 5   Publisher     16327 non-null  category
 6   NA_Sales      16327 non-null  float16 
 7   EU_Sales      16327 non-null  float16 
 8   JP_Sales      16327 non-null  float16 
 9   Other_Sales   16327 non-null  float16 
 10  Global_Sales  16327 non-null  float64 
dtypes: category(3), float16(4), float64(1), int16(1), int64(1), object(1)
memory usage: 3.3 MB


In [102]:
games_copy.Global_Sales = games_copy.Global_Sales.astype('float16')

In [103]:
games_copy.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
Index: 16327 entries, Wii Sports to Spirits & Spells
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype   
---  ------        --------------  -----   
 0   Rank          16327 non-null  int64   
 1   Name          16327 non-null  object  
 2   Platform      16327 non-null  category
 3   Year          16327 non-null  int16   
 4   Genre         16327 non-null  category
 5   Publisher     16327 non-null  category
 6   NA_Sales      16327 non-null  float16 
 7   EU_Sales      16327 non-null  float16 
 8   JP_Sales      16327 non-null  float16 
 9   Other_Sales   16327 non-null  float16 
 10  Global_Sales  16327 non-null  float16 
dtypes: category(3), float16(5), int16(1), int64(1), object(1)
memory usage: 3.2 MB


In [104]:
games.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16598 entries, 0 to 16597
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Rank          16598 non-null  int64  
 1   Name          16598 non-null  object 
 2   Platform      16598 non-null  object 
 3   Year          16327 non-null  float64
 4   Genre         16598 non-null  object 
 5   Publisher     16540 non-null  object 
 6   NA_Sales      16598 non-null  float64
 7   EU_Sales      16598 non-null  float64
 8   JP_Sales      16598 non-null  float64
 9   Other_Sales   16598 non-null  float64
 10  Global_Sales  16598 non-null  float64
dtypes: float64(6), int64(1), object(4)
memory usage: 5.2 MB


as we can see above, we reduced the size of this dataset was reduced from 5.2 MB to 3.2 MB

# References:
-https://www.kaggle.com/gregorut/videogamesales